### PR TITLE
[MM-31934] Data retention settings page

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
This is the landing page for data retention.

It contains:
- A table for the global retention policy.
- A table for custom policies and the ability to delete a policy
- The job table and a button to kickstart a job
- An edit button just above the job table to edit the job time
- The page is behind a feature flag, you will need to enable `CustomDataRetentionEnabled`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-31934
https://www.figma.com/file/PdZPkf36lNU6ZO41JlEjKd/Team-%26-Channel-Data-Retention?node-id=0%3A1

#### Related Pull Requests
There are a number of PRs open with the same changes as here, they are almost merged:
- https://github.com/mattermost/mattermost-webapp/pull/7467
- https://github.com/mattermost/mattermost-webapp/pull/7743
- https://github.com/mattermost/mattermost-webapp/pull/7816

Due to the above PRs you can skip over:
- components/admin_console/data_grid/
- components/admin_console/jobs/

```release-note
New data retention landing page hidden behind a feature flag. It shows the global retention settings, granular retention policies and the job table.
```

### CODE REVIEW

**Review by sbishel — CHANGES_REQUESTED** [review]
Just a couple minor things. Looks good.

File: components/admin_console/data_retention_settings/data_retention_settings.tsx [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+            for (let i = 0; i < minuteIntervals.length; i++) {
+                options.push({label: `${hourLabel}:${minuteIntervals[i]}${timeOfDay}`, value: `${hourValue}:${minuteIntervals[i]}`});
+            }
+        }
+
+        return options;
+    }
+
+    changeJobTimeConfig = async (value: string) => {
+        const newConfig = JSON.parse(JSON.stringify(this.props.config));
```
Comment (anurag6713) [inline components/admin_console/data_retention_settings/data_retention_settings.tsx:350]: Non blocking question. Can we have a util method for deep extend? Example Deep Extend section in http://youmightnotneedjquery.com.

File: components/admin_console/data_retention_settings/data_retention_settings.tsx [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+            if (hourLabel < 10) {
+                hourValue = `0${hourValue}`;
+            }
+            if (hourLabel > 12) {
+                hourLabel -= 12;
+            }
+            for (let i = 0; i < minuteIntervals.length; i++) {
+                options.push({label: `${hourLabel}:${minuteIntervals[i]}${timeOfDay}`, value: `${hourValue}:${minuteIntervals[i]}`});
+            }
+        }
```
[the remark this answers is not in the document]
Comment (BenCookie95) [reply components/admin_console/data_retention_settings/data_retention_settings.tsx]: Removed that first loop and memoized the result. Should only run once now

File: components/admin_console/data_retention_settings/data_retention_settings.tsx (hunk 1 above)
Comment (BenCookie95) [reply components/admin_console/data_retention_settings/data_retention_settings.tsx:327]: Yeah I could add something like that in the future but this is ok for now I think

**Review by sbishel — APPROVED** [review]
NIce, thanks for the changes.

**Review by anurag6713 — APPROVED** [review]

**Review by furqanmlk — COMMENTED** [review]
image

**FIGMA**
image

**Comment by BenCookie95** [comment]
@furqanmlk I have fixed the titles!

**Review by furqanmlk — APPROVED** [review]
Thanks @BenCookie95 for updating.

**Pull request merged**